Tidy up scripts

Baschdl opened the pull request:
@bjrne @maltezacharias What do you think? Do you have an idea for a cleaner solution?

Baschdl committed:
Move backup.sh to scripts folder

Baschdl committed:
Activate already installed django-extensions

Baschdl committed:
Use db agnostic reset and add possibility to run commands in docker

maltezacharias commented on scripts/delete_db_and_setup.sh:
```diff
@@ -1,6 +1,10 @@
-echo "Please enter your desired email and password:"
-python3 backend/manage.py createsuperuser  --username admin
-python3 backend/manage.py loaddata backend/cms_export.json
+#!/bin/bash
```
```suggestion
#!/bin/bash -e
```
(Will exit when a command fails)

maltezacharias commented on scripts/delete_db_and_setup.sh:
```diff
@@ -1,6 +1,10 @@
+#!/bin/bash
+# To run script in Docker call for example "delete_db_and_setup.sh 'docker-compose exec backend'"
+EXEC_ENV="$*"
+
```
It would be nice if you could run it from anywhere, but I don't know how to do that..
I tried this:
```suggestion
DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )" && cd $DIR/../
```

but then it would be a bit unexpected that you need to provide a docker-compose command taking into account the directory switch. On the other hand you wouldn't need to care about where you are when you call that script, it would always be `docker-compose -f docker-compose.yml -f docker-compose.prod.yml`

What do you think @Baschdl ?

Baschdl replied:
I think we would also have to specify the path of `backend/cms_export.json` dynamically. Maybe that's a bit too much engineering?

maltezacharias replied:
Probably.. The suggestion I wrote switches to the project root dir reliably no matter from where you call the script (copied from StackOverflow, there is an error condition somewhere with symlinks that I forgot, but as you said... Didn't want to over-engineer. I wouldn't invest more time I think

Baschdl replied:
Not sure what to do here, I really like the idea but it may be a bit of an unexpected behavior for someone using it. We should probably maximize for the understandability for a new user

bjrne replied:
Right now it needs to be called with `bash ../scripts/delete_and_setup_db.py` right?

Baschdl replied:
Yes, `cd backend && bash ../scripts/delete_and_setup_db.py` if you don't use docker

maltezacharias replied:
With the proposed change it should work from everywhere.

Baschdl replied:
> would be a bit unexpected that you need to provide a docker-compose command taking into account the directory switch

@maltezacharias I'm not sure what you mean with this. It's always relative to `backend` if you use docker, right?

maltezacharias replied:
My suggested line always checks where the script is located (the file itself) and then changes the workdir to the directory above. So you would always execute your commands from the repo root. That would mean you can either call it in one of the following ways:

```
# Using an absolute path from anywhere
/home/user/match4everything/scripts/delete_db_and_setup.sh  # for native django
/home/user/match4everything/scripts/delete_db_and_setup.sh docker-compose exec backend # for dev docker
/home/user/match4everything/scripts/delete_db_and_setup.sh docker-compose -f docker-compose.yml -f docker-compose.prod.yml exec backend # for dev docker

# From scripts dir
cd scripts
./delete_db_and_setup.sh  # for native django
./delete_db_and_setup.sh docker-compose exec backend # for dev docker
./delete_db_and_setup.sh docker-compose -f docker-compose.yml -f docker-compose.prod.yml exec backend # for dev docker

# from repo root
./scripts/delete_db_and_setup.sh  # for native django
./scripts/delete_db_and_setup.sh docker-compose exec backend # for dev docker
./scripts/delete_db_and_setup.sh docker-compose -f docker-compose.yml -f docker-compose.prod.yml exec backend # for dev docker
```

All of these should work equally (If I didn't make a typo)

maltezacharias commented on scripts/delete_db_and_setup.sh:
```diff
@@ -1,6 +1,10 @@
+# To run script in Docker call for example "delete_db_and_setup.sh 'docker-compose exec backend'"
+EXEC_ENV="$*"
+
+$EXEC_ENV python3 manage.py reset_db
```
Neat!

maltezacharias commented:
Please review above suggestions. The PR looks fine to me with or without them.

Baschdl committed:
Exit when commands fail

Co-authored-by: Malte Zacharias <mz-basis+github@asenwelt.de>

bjrne reviewed (commented):
Is the change in permissions intended?
100644 → 100755

bjrne committed:
Remove whitespace

bjrne committed:
Add script for devs

bjrne committed:
Give fakeusers shorter passwords

bjrne committed:
Secure currently unused path of user generation

bjrne commented:
> Do you have an idea for a cleaner solution?

I don't :sweat_smile:

Baschdl commented:
> Is the change in permissions intended?

Yes, it makes it executable

Baschdl commented on backend/apps/matching/management/commands/createfakeusers.py:
```diff
@@ -17,7 +17,7 @@ class Command(BaseCommand):
     help = (  # noqa
         "Populates the database with fake users or deletes them."
-        "Every user has their email as their password."
+        "Every user has type (A|B) as their password."
```
@bjrne But why? I find it a lot more straightforward to just copy down the mail address.

bjrne replied:
Copy is two clicks plus keystroke, this is one click plus keystroke ;) I you don't like it reverse it, I have several logins saved as credentials in the browser.

Baschdl replied:
It's "Helper" or "Institution" now, so the mail is easier

Baschdl commented on backend/apps/matching/models/user.py:
```diff
@@ -34,7 +35,7 @@ def new(email, pwd=None, **kwargs):
         if pwd is not None:
             user.set_password(pwd)
         else:
-            user.set_password(email)
+            user.set_password(get_random_string(length=32))
```
I think we should switch to `def new(email, pwd, **kwargs)` and don't have any process of setting an automatic password.

bjrne replied:
Sure

Baschdl commented on scripts/clean_db_and_setup_with_fake_users.sh:
```diff
@@ -0,0 +1,14 @@
+EXEC_ENV="$*"
+DJANGO_SUPERUSER_PASSWORD="admin"
+
+$EXEC_ENV echo "DO NOT USE THIS SCRIPT IN PRODUCTION"
```
Why don't we leave it general enough for production and development? I have actively decided against including the `createfakeusers` etc.

bjrne replied:
Well then just delete all my commits. I have an alias for this, but then you could just make the script docker only. If you need to enter this many things and also create fake users manually, I won't use it for development, and I suspect other wont aswell.

Baschdl replied:
Merged both scripts

bjrne committed:
Merge branch 'staging' into tidy_up_scripts

Baschdl committed:
Revert "Give fakeusers shorter passwords"

This reverts commit b86cf5d988f39819ffff599bb04365c2ff2cfe92.

Baschdl committed:
Always require password

Baschdl committed:
Merge both scripts

bjrne commented on scripts/delete_db_and_setup.sh:
```diff
@@ -1,6 +1,29 @@
+  exit 1
+fi
+if [ $MODE == 'DEV' ]; then
+  echo "DO NOT USE THIS SCRIPT IN PRODUCTION"
```
Maybe this hint is misleading when there is a production version in the same script? But I don't have a better idea.

bjrne reviewed (approved):
This needs some documentation, but then it's ready.

Baschdl commented on scripts/delete_db_and_setup.sh:
```diff
@@ -1,6 +1,29 @@
+  exit 1
+fi
+if [ $MODE == 'DEV' ]; then
+  echo "DO NOT USE THIS SCRIPT IN PRODUCTION"
```
```suggestion
  echo "DO NOT USE THE 'DEV' MODE IN PRODUCTION"
```
@bjrne How about this?

bjrne replied:
Sounds funny, but better than everything I've come up with.

Baschdl committed:
Update warning message

Baschdl committed:
Add documentation

bjrne commented on scripts/delete_db_and_setup.sh:
```diff
@@ -21,7 +21,7 @@ if [ $MODE == 'PROD' ]; then
   echo "Please enter your desired email and password."
   $EXEC_ENV python3 manage.py createsuperuser --username admin
 else
-  $EXEC_ENV python3 manage.py createsuperuser --username admin --email admin@examplee.com --no-input
```
This was actually intentional in order to preserve the superuser when deleting all fakeusers with the script. But I can live with it both ways since I rarely use the fakeuser-deletion.

Baschdl replied:
We can add it again. Without documentation this looks like a typo 😬

bjrne committed:
Merge branch 'staging' into tidy_up_scripts

bjrne reviewed (approved):
Let's merge this? @Baschdl

Baschdl commented:
@bjrne I would wait for the answer from @maltezacharias but I'm also missing this script in the other branches

maltezacharias commented:
> @bjrne I would wait for the answer from @maltezacharias but I'm also missing this script in the other branches

please ping on telegram if the answer is not sufficient, don't want to keep you waiting

maltezacharias committed:
Merge branch 'staging' into tidy_up_scripts

maltezacharias committed:
Update scripts/delete_db_and_setup.sh

bjrne committed:
Add help message when no argument is supplied

bjrne committed:
Use backend prefix since script now runs from root dir

bjrne committed:
Skip system checks in all statements except for last one

They were executed multiple times and prevented important output to be read properly,
like the "Creating account with username \"admin\". " hint.
